File: android/src/main/java/top/wj/zk_vpn/ZkVpnPlugin.java
package top.wj.zk_vpn;

import android.app.Activity;
import android.app.Application;
import android.util.Log;

import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodChannel;

public class ZkVpnPlugin implements FlutterPlugin, ActivityAware {
    private static final String TAG = "ZkVpnPlugin";
    private MethodChannel methodChannel;
    private EventChannel eventChannel;
    private FlutterPluginBinding pluginBinding;
    private Application application;
    private Activity activity;

    @Override
    public void onAttachedToEngine(@NonNull FlutterPlugin.FlutterPluginBinding binding) {
        Log.i(TAG, "onAttachedToEngine: ");
        this.pluginBinding = binding;
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPlugin.FlutterPluginBinding binding) {
        Log.i(TAG, "onDetachedFromEngine: ");
        this.pluginBinding = null;
    }

    @Override
    public void onAttachedToActivity(@NonNull ActivityPluginBinding binding) {
        Log.i(TAG, "onAttachedToActivity: ");
        if (pluginBinding != null) {
            setup(pluginBinding, binding);
        }
    }

    @Override
    public void onDetachedFromActivityForConfigChanges() {
        Log.i(TAG, "onDetachedFromActivityForConfigChanges: ");
        onDetachedFromActivity();
    }

    @Override
    public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {
        Log.i(TAG, "onReattachedToActivityForConfigChanges: ");
        onAttachedToActivity(binding);
    }

    @Override
    public void onDetachedFromActivity() {
        Log.i(TAG, "onDetachedFromActivity: ");
        teardown();
    }

    private void setup(FlutterPluginBinding fBin, ActivityPluginBinding aBin) {
        methodChannel = new MethodChannel(fBin.getBinaryMessenger(), "zk_vpn");
//        eventChannel = new EventChannel(fBin.getBinaryMessenger(), "zk_vpn_event");

        this.application = (Application) fBin.getApplicationContext();
        this.activity = aBin.getActivity();

        PluginMethodCall pluginMethodCall = new PluginMethodCall(this.activity);
        PluginStream pluginStream = new PluginStream(this.activity);

        methodChannel.setMethodCallHandler(pluginMethodCall);
        eventChannel.setStreamHandler(pluginStream);
    }

    private void teardown() {
        methodChannel.setMethodCallHandler(null);
//        eventChannel.setStreamHandler(null);

        application = null;
        activity = null;

        methodChannel = null;
        eventChannel = null;
    }
}

File: android/src/main/kotlin/top/wiz/zk_vpn/ZkVpnPlugin.kt
package top.wiz.zk_vpn

import android.app.Activity
import android.app.Application
import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.PluginRegistry

class ZkVpnPlugin : FlutterPlugin, ActivityAware {

    private var methodChannel: MethodChannel? = null
    private var pluginBinding: FlutterPlugin.FlutterPluginBinding? = null
    private var application: Application? = null
    private var activity: Activity? = null

    companion object {
        private const val TAG = "ZkVpnPlugin"

        @Deprecated("only use flutter v1")
        fun registerWith(registrar: PluginRegistry.Registrar) {
            if (registrar.activity() == null) return
            val plugin = ZkVpnPlugin()
            val application = registrar.activeContext().applicationContext as Application
            plugin.setup(registrar.messenger(), application, registrar.activity())
        }
    }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        Log.i(TAG, "onAttachedToEngine: ")
        this.pluginBinding = binding
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        Log.i(TAG, "onDetachedFromEngine: ")
        this.pluginBinding = null
    }

    override fun onDetachedFromActivity() {
        Log.i(TAG, "onDetachedFromActivity: ")
        teardown()
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        Log.i(TAG, "onReattachedToActivityForConfigChanges: ")
        onAttachedToActivity(binding)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        Log.i(TAG, "onAttachedToActivity: ")
        pluginBinding?.let {
            setup(it.binaryMessenger, it.applicationContext as Application, binding.activity)
        }
    }

    override fun onDetachedFromActivityForConfigChanges() {
        Log.i(TAG, "onDetachedFromActivityForConfigChanges: ")
        onDetachedFromActivity()
    }

    private fun setup(messenger: BinaryMessenger, application: Application, activity: Activity) {
        methodChannel = MethodChannel(messenger, "zk_vpn")

        this.application = application
        this.activity = activity

        val handler = ZkVpnMethodCallHandler(activity)
        methodChannel?.setMethodCallHandler(handler)
    }

    private fun teardown() {
        methodChannel?.setMethodCallHandler(null)
        application = null
        activity = null
        methodChannel = null
    }
}
